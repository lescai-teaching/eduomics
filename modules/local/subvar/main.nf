process SUBVAR {
    tag "$meta.id"
    label 'process_low'
    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/htslib:1.20--h5efdd21_2' :
        'biocontainers/htslib:1.20--h5efdd21_2'}"

    input:
    tuple val(meta), path(vcf)
    path bed

    output:
    tuple val(meta), path('*_benign.vcf.gz')        , optional: true, emit: benign_vcf
    tuple val(meta), path('*_pathogenic.vcf.gz')    , optional: true, emit: pathogenic_vcf
    tuple val(meta), path('*_selected.vcf')         , optional: true, emit: selected_vcf
    path 'versions.yml'                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    tabix -p vcf ${vcf}
    tabix -h -R ${bed} ${vcf} | bgzip -c > ${prefix}_ontarget.vcf.gz
    zcat ${prefix}_ontarget.vcf.gz | grep -e "^#" -e "Benign" -e "_benign" | bgzip -c > ${prefix}_benign.vcf.gz
    zcat ${prefix}_ontarget.vcf.gz | grep -e "^#" -e "Pathogenic" -e "_pathogenic" | bgzip -c > ${prefix}_pathogenic.vcf.gz

    # Additional filtering steps
    zcat ${prefix}_ontarget.vcf.gz | grep -v "no_assertion_criteria_provided" | grep -e "^#" -e "reviewed_by_expert_panel" > ${prefix}_selected.vcf
    zcat ${prefix}_ontarget.vcf.gz | grep -v "^#" | grep -e "multiple_submitters" | grep "nonsense" >> ${prefix}_selected.vcf



    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        tabix: \$(tabix --version 2>&1 | sed 's/^.*tabix //')
        bgzip: \$(bgzip --version 2>&1 | sed 's/^.*bgzip //')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    touch ${prefix}_benign.vcf.gz
    touch ${prefix}_pathogenic.vcf.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        tabix: \$(echo "1.20")
        bgzip: \$(echo "1.20")
    END_VERSIONS
    """
}
