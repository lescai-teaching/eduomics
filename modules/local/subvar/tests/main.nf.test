nextflow_process {

    name "Test Process SUBVAR"
    script "../main.nf"
    process "SUBVAR"

    tag "modules"
    tag "modules_nfcore"
    tag "subvar"

    test("homo_sapiens - vcf") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ], // meta map
                    file('/Users/bagordo/Downloads/clinvar_20250317_chr22q11-21.vcf.gz', checkIfExists: true)
                ]
                input[1] = file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/main/dna/Twist_Exome_Target2_chr22q11-21_GRCh38_50pad.bed', checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path(process.out.benign_vcf.get(0).get(1)).exists() },
                { assert path(process.out.pathogenic_vcf.get(0).get(1)).exists() },
                { assert path(process.out.versions.get(0)).exists() }
            )
        }

    }

    test("homo_sapiens - vcf - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ], // meta map
                    file("/Users/bagordo/Downloads/clinvar_20250317_chr22q11-21.vcf.gz", checkIfExists: true)
                ]
                input[1] = file("https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/main/dna/Twist_Exome_Target2_chr22q11-21_GRCh38_50pad.bed", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path(process.out.benign_vcf.get(0).get(1)).exists() },
                { assert path(process.out.pathogenic_vcf.get(0).get(1)).exists() },
                { assert path(process.out.versions.get(0)).exists() }
            )
        }
    }
}
