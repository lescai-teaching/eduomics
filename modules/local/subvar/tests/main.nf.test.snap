{
    "homo_sapiens - vcf": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "bam": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-24T12:26:25.216801"
    },
    "homo_sapiens - vcf - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "bam": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-24T12:26:46.228221"
    }
}